### DISCUSSION TITLE
RFC: Safer Capabilities

### DISCUSSION THREAD

**Discussion by eudoxia0** [austral/austral#438] [Ideas]
# Summary

Austral's capability security system is not secure enough. This proposal might work to make it more secure.

# How Capabilities Work Currently

Capability security is not a "first class" feature of Austral. It's mostly a consequence of how linear types work. The only thing that has to be built into the language to enable it is the `RootCapability` type. Values of this type cannot be constructed manually, and the only way to acquire it is as a parameter to the program entrypoint function.

Currently, capabilities mostly act as type-level permission slips for performing side effects. For example:

```austral
module Silo is
    type LaunchCapability: Linear;

    generic [R: Region]
    function launchMissiles(cap: &[LaunchCapability, R]): Unit;

    generic [R: Region]
    function acquire(ref: &[RootCapability, R]): LaunchCapability;

    function release(cap: LaunchCapability): Unit;
end module.
```

The `Silo` module has a `LaunchCapability` type, representing the capability for launching the missiles. `acquire` returns an instance of `LaunchCapability` from a reference to the `RootCapability`, and `release` is the consumer of `LaunchCapability`

You'd use it like so:

```austral
let cap: LaunchCapability := acquire(&root);
launchMissiles(&cap);
release(cap);
```

And the implementation of this module would be something like this:

```austral
module body Silo is
    pragma Unsafe_Module;

    record LaunchCapability: Linear
    end;

    function unsafeLaunch(): Unit
        -- Binding to a foreign C function.
        pragma Foreign_Import(External_Name => "launch_missiles");
    end;

    generic [R: Region]
    function launchMissiles(cap: &[LaunchCapability, R]): Unit is
        unsafeLaunch();
        return nil;
    end;

    generic [R: Region]
    function acquire(ref: &[RootCapability, R]): LaunchCapability is
        return LaunchCapability();
    end;

    function release(cap: LaunchCapability): Unit;
        let {} := cap;
        return nil;
    end.
end module body.
```

The thing to notice here is that `LaunchCapability` is just an empty record. It exists to prove the client at some point had a reference to the `RootCapability`, and this permission slip can then be used to call `launchMissiles`.

In short: the module has a linearly-typed, capability secure interface; enclosing an unsafe implementation.

# Problem

Why is this a problem? Well, the point of capability security is, mostly, to make [supply chain attacks][sup] harder.

[sup]: https://en.wikipedia.org/wiki/Supply_chain_attack

But doesn't the FFI completely defeat the point? Well, yes, and no. Every language (possibly excepting [Vale][vale]) needs an unsafe escape hatch at the FFI boundary. This is as true for Haskell as it is for Python and Rust. The difference is essentially one of _scope_. In Python, _any_ line of code could be malware and has to be audited. In Austral, the idea is that unsafe modules expose safe interfaces, and so only unsafe modules have to be audited.

[vale]: https://vale.dev/

So capability security doesn't _eliminate_ supply chain attacks, it only reduces the amount of work you have to perform to certify a set of dependencies is more or less safe.

But consider this scenario: a Python package in PyPI is compromised and replaced with malware. Thousands of devs download it transitively and it infects their computers. If this happened with Austral, would it be any different? An Austral package is replaced with malware, that uses the `Unsafe_Module` pragma to create its own interface to the underlying C functions and completely bypass the capability hierarchy in the (future) standard library.

Users would be upset by this. It seems "capability security" is not security at all. Having to perform an audit each time you download a package is a very onerous demand to place on programmers.

Can we do better?

# Solution

tl;dr: in this approach, auditing is unnecessary, and it is _impossible_ to surreptitiously use the FFI.

The basic idea is to create a new capability (called `ForeignCapability` or some such) that has to be passed to the unsafe functions in the `Austral.Memory` module, and to all FFI functions, and get rid of `Unsafe_Pragma`.

Summary of the changes:

1. Get rid of the `Unsafe_Module` pragma.
1. Any module can import `Austral.Memory`.
1. Any module can define a foreign function for import.
1. There is a new type `ForeignCapability: Linear`, along with functions to get
   it from a reference to the `RootCapability` and to release it:
   1. `function acquireForeign(cap: &[RootCapability, R]): ForeignCapability`
   1. `function releaseForeign(cap: ForeignCapability): Unit;`
1. Every unsafe function in `Austral.Memory` takes a `&[ForeignCapability, R]`
   reference as its first argument.
1. Every foreign function also takes a `&[ForeignCapability, R]` reference as
   its first argument. So if you write a binding to `foo(int)` you'd have to
   call it like `foo(&cap, 10)`.

The above example would be updated like this:

```austral
module Silo is
    type LaunchCapability: Linear;

    generic [R: Region]
    function launchMissiles(cap: &[LaunchCapability, R]): Unit;

    generic [R: Region]
    function acquire(ref: &[RootCapability, R]): LaunchCapability;

    function release(cap: LaunchCapability): Unit;
end module.
```

And the implementation:

```austral
module body Silo is
    record LaunchCapability: Linear
        foreign: ForeignCapability;
    end;

    generic [R: Region]
    function unsafeLaunch(cap: &[ForeignCapability, R): Unit
        pragma Foreign_Import(External_Name => "launch_missiles");
    end;

    generic [R: Region]
    function launchMissiles(cap: &[LaunchCapability, R]): Unit is
        unsafeLaunch(cap->foreign);
        return nil;
    end;

    generic [R: Region]
    function acquire(ref: &[RootCapability, R]): LaunchCapability is
        let foreign: ForeignCapability := acquireForeign(ref);
        return LaunchCapability(foreign);
    end;

    function release(cap: LaunchCapability): Unit;
        let { foreign: ForeignCapability } := cap;
        releaseForeign(foreign);
        return nil;
    end.
end module body.
```

Instead of being an empty record, `LaunchCapability` is now a wrapper that hides a `ForeignCapability`, and passes it to the FFI function.

# Problems

1. What about foreign functions that _don't_ have side effects, like e.g. the    mathematical functions in `libm`? It would be absurd to require users to pass a `MathCapability` to call a foreign version of the `sin` function. Maybe it's possible to whitelist certain bindings within the standard library, but every custom library needs to pass a `ForeignCapability` to access even pure foreign functions.

**Comment by crabmusket** [comment]
Re-reading this and the similar ideas [here](https://borretti.me/article/how-capabilities-work-austral#strict), I wanted to ask: are FFI and raw memory separate classes of unsafety? I haven't thought through this in great detail yet, but my gut says that being able to allocate some bytes and play with them should be classified as lower-risk than calling arbitrary C code.

The Austral.Memory module currently exposes a `Pointer` type which is `Free` which allows leaking memory. Would it be beneficial to also introduce a "safe" raw memory module with linear types and checked access to memory? Analogous to a safe filesystem API, where you can't write to files outside the tree you have been given the capability to access.

EDIT: "safe memory access" sounds like the basis for Vale's wasm ffi? It's "safe" to play with a sandboxed chunk of memory however you want to.

**Comment by eudoxia0** [comment]
>are FFI and raw memory separate classes of unsafety?

I think of it less as separate classes and more about granularity of permissions. Sure, calling arbitrary C code is the higher permission. But sometimes you just want to allow a function to, say, offset a pointer and read from memory, but not write from memory.

>The Austral.Memory module currently exposes a Pointer type which is Free which allows leaking memory. Would it be beneficial to also introduce a "safe" raw memory module with linear types and checked access to memory? Analogous to a safe filesystem API, where you can't write to files outside the tree you have been given the capability to access.

Yes, so the standard library has a `Box` type (analogous to Rust's `[Box](https://doc.rust-lang.org/std/boxed/struct.Box.html)`) that is the linear, null-safe etc. version of a pointer to a single object. The `Buffer` type is a linear, heap-allocated resizable array and that's the safe version of a pointer to multiple objects.

So you have two categories:

1. Unsafe, `Free` pointers.
2. Safe, linear standard library containers.

There isn't a third category in the middle and I'm not sure what would exist in that category.

**Comment by bwburnsides** [comment]
Looks like the discussion didn't go very far here. Was any additional thought given to this idea?
